src/cursor/cur_table.c

```diff
@@ -837,18 +866,29 @@ __curtable_largest_key(WT_CURSOR *cursor)
 static int
 __curtable_bound(WT_CURSOR *cursor, const char *config)
 {
-    WT_CURSOR *primary;
+    WT_CURSOR **cp, *primary;
+    WT_CURSOR_BOUNDS_STATE saved_bounds;
     WT_CURSOR_TABLE *ctable;
     WT_DECL_RET;
     WT_SESSION_IMPL *session;
+    u_int i, j;
 
+    i = 0;
     ctable = (WT_CURSOR_TABLE *)cursor;
+    primary = *ctable->cg_cursors;
     JOINABLE_CURSOR_API_CALL(cursor, session, bound, NULL);
 
-    /* Grab the primary cursor and call bound function. */
-    primary = *ctable->cg_cursors;
-    WT_ERR(primary->bound(primary, config));
+    __wt_cursor_bounds_save(primary, &saved_bounds);
+
+    /* Call bound function on all column groups. */
+    for (i = 0, cp = ctable->cg_cursors; i < WT_COLGROUPS(ctable->table); i++, cp++)
+        WT_ERR((*cp)->bound(*cp, config));
 err:
+    /* If applying bounds fails on one colgroup, reset all of them for consistency. */
+    if (ret != 0) {
+        for (j = 0, cp = ctable->cg_cursors; j <= i; j++, cp++)
```

Comment:
Personally I think this optimisation is increases the performance that doesn't really benefit too much and personally prefer that we just work from 0 to the end of all colgroups for easier read purposes. what are your thoughts @luke-pearson ?

Reply:
This is a tricky one, on the one hand there is a lower probability of bugs. On the other hand the logic here looks correct to me and we would do a lot less operations overall. I'm happy to go with the i = 0 loop. Reset all the things loop.